docs/internal/kubernetes_jobs.md

```diff
@@ -0,0 +1,45 @@
+
+Indexed completion mode supports partitioned yaml files generated by `generate-table-partitions` in Data Validation, if each worker process ran only the yaml file corresponding to its index. I have an introduced an optional parameter `--kube-completions` or `-kc`. When this flag is used with `data-validation configs run -cdir` with a config directory and the container runs in indexed jobs mode, each container only processes the specific validation yaml file corresponding to its index. If the flag is used with `data-validation configs run -cdir` with a config directory and DVT is not running in indexed jobs mode, a warning is issued. In other uses of `configs` such as `configs list` and `configs run -c`, this flag is ignored.
+### IAM Permissions
+This section is yet to be written and can be updated with the required IAM permissions. This section probably belongs in the external facing documentation.
+### Passing database connection parameters
+DVT database connection parameters are saved in `$HOME/.config/google-pso-data-validator` directory with passwords in raw text. With Kubernetes, DVT cannot depend on `.config` directory holding the connections unless they are baked into the image (or mounted as a volume - see Cloud Run limitation below) - which will require each customer to modify the container image we provide. A better approach (for regular and containerized DVT) would be use the (GCP) Secret Manager and retrieve connection credentials as a JSON object when we connect to the database. DVT currently uses the Secret Manager for retrieving secrets and stores them in the `.config` directory when the connections are added. This seems rather odd as it defeats the main purposes of using the secret manager - security and password rotation. 
+
+I am proposing a simple command line change to resolve this issue. Whenever a connection parameter is specified, allow the user to optionally specify a secret manager (provider, project-id). If a secret manager is specified, then DVT retrieves the connection information directly from the secret manager at the time of creating the connection. This is the recommended approach to handle secrets as opposed to mounting secrets as volumes. Cloud Run also has a limitation that multiple secrets [cannot be mounted with the same path](https://cloud.google.com/run/docs/configuring/services/secrets#disallowed_paths_and_limitations). Since DVT requires connections to two different databases with the connection info being mounted in the same directory, i.e. `$HOME/.config/google-pso-data-validator`, DVT cannot run within Cloud Run without this change. With this change, DVT can be run in a container in Cloud Run or Kubernetes fetching the connection information from the GCP Secret Manager. 
```

Comment:
Introducing the secret-manager-project/type for every validation command removes the de-coupling of connection configuration and validation configuration. I understand the issue here is storing connection configs locally, which would be inconvenient in Kubernetes. 
But why can't we use GCS connections + Secret Manager to get the correct config at runtime? We would only need to set the PSO_CONFIG_DIR env variable to enable GCS connections.

Reply:
Neha,

I had not thought of that as an option initially. There are some challenges:

1. The directory structure under PSO_DV_CONFIG is two directories, one `connections` with connection information and another directory `validations` with yaml files. When we look for yaml files we only look for yaml files in the current directory - so in this case under `validations`. The way `generate-table-partitions` works, it creates yaml files under the _config-dir/<schema_name>.<table_name>_. Therefore PSO_DV_CONFIG cannot be directly used - we would need to copy files after running `generate-table-partitions`.
2. We have an [open security issue](https://github.com/GoogleCloudPlatform/professional-services-data-validator/issues/756) with using GCS (or file system) for connection information.
3. Database passwords and connection information are secrets and the proposed implementation is simpler than the current implementation which requires multiple secrets per connection.

Hope that helps.

Sundar Mudupalli

Reply:
Yeah - Issue #756 which you mentioned involves de-coupling `connections/` and `validations/`. I think we need to address this issue before we move forward with Kubernetes or other distributed validations. This addresses your first point as well - theoretically `generate-table-partitions -cdir my-table/validations/` should drop the YAML in the right directory.

I agree with your proposed change to make the entire connection JSON a secret rather than each field in the JSON - we should move forward with that. I just think we need to reconsider adding secret manager flags to each command and instead use GCS based connections, even if we need to address Issue 756 first.